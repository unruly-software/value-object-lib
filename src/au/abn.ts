import { ValueObject } from '@unruly-software/value-object'
import z from 'zod'

const WEIGHTS = [10, 1, 3, 5, 7, 9, 11, 13, 15, 17, 19]

export class ABN extends ValueObject.define({
  id: '@unruly-software/value-object-lib/ABN',
  schema: () =>
    z
      .string({
        message: 'Invalid ABN. Must be a text value.',
      })
      .trim()
      .transform((initial, ctx) => {
        let cleanValue: string = ''
        const digits: number[] = []

        for (const char of initial) {
          if (char === '-' || char === ' ') {
            continue
          }
          const number = parseInt(char, 10)
          if (Number.isNaN(number)) {
            ctx.addIssue({
              code: z.ZodIssueCode.custom,
              message: 'Invalid ABN. Must be a text value.',
            })
            return z.NEVER
          }
          cleanValue += char
          digits.push(number)
        }

        if (digits.length !== 11) {
          ctx.addIssue({
            code: z.ZodIssueCode.custom,
            message: 'Invalid ABN format. Must be 11 digits long.',
          })
          return z.NEVER
        }

        digits[0] -= 1
        const checksum = digits.reduce((sum, number, index) => {
          const weightedNumber = number * WEIGHTS[index]
          return sum + weightedNumber
        }, 0)

        if (checksum % 89 !== 0) {
          ctx.addIssue({
            code: z.ZodIssueCode.custom,
            message: 'Invalid ABN checksum.',
          })
          return z.NEVER
        }

        return cleanValue
      }),
}) {
  static fakeInstance() {
    return ABN.fromJSON('76 861 743 447')
  }

  /**
   * Australian ABNs are typically formatted as "xx xxx xxx xxx" when
   * displayed.
   */
  toStandardFormat() {
    const parts = this.props.split('')
    return [
      parts.splice(0, 2).join(''),
      parts.splice(0, 3).join(''),
      parts.splice(0, 3).join(''),
      parts.splice(0, 3).join(''),
    ].join(' ')
  }
}
