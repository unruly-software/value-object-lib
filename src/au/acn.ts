import { ValueObject } from '@unruly-software/value-object'
import z from 'zod'

const WEIGHTS = [8, 7, 6, 5, 4, 3, 2, 1]

export class ACN extends ValueObject.define({
  id: '@unruly-software/value-object-lib/ACN',
  schema: () =>
    z
      .string({
        message: 'Invalid ACN. Must be a text value.',
      })
      .trim()
      .transform((initial, ctx) => {
        let cleanValue: string = ''
        const digits: number[] = []

        for (const char of initial) {
          if (char === '-' || char === ' ') {
            continue
          }
          const number = parseInt(char, 10)
          if (Number.isNaN(number)) {
            ctx.addIssue({
              code: z.ZodIssueCode.custom,
              message: 'Invalid ACN. Must be a text value.',
            })
            return z.NEVER
          }
          cleanValue += char
          digits.push(number)
        }

        if (digits.length !== 9) {
          ctx.addIssue({
            code: z.ZodIssueCode.custom,
            message: 'Invalid ACN format. Must be 11 digits long.',
          })
          return z.NEVER
        }

        const first8Numbers = digits.slice(0, -1)
        const numbersSum = first8Numbers.reduce((sum, number, index) => {
          const weightedNumber = number * WEIGHTS[index]
          return sum + weightedNumber
        }, 0)
        const remainder = numbersSum % 10
        const complement = 10 - remainder
        const checkDigit = complement === 10 ? 0 : complement

        if (digits[8] !== checkDigit) {
          ctx.addIssue({
            code: z.ZodIssueCode.custom,
            message: 'Invalid ACN checksum.',
          })
          return z.NEVER
        }

        return cleanValue
      }),
}) {
  static fakeInstance() {
    return ACN.fromJSON('000 000 019')
  }

  /**
   * Australian ACNs are typically formatted as "xxx xxx xxx"
   * when displayed.
   */
  toStandardFormat() {
    const parts = this.props.split('')
    return [
      parts.splice(0, 3).join(''),
      parts.splice(0, 3).join(''),
      parts.splice(0, 3).join(''),
    ].join(' ')
  }
}
