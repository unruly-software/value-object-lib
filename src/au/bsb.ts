import { ValueObject } from '@unruly-software/value-object'
import z from 'zod'

export class BSB extends ValueObject.define({
  id: '@unruly-software/value-object-lib/BSB',
  schema: () =>
    z
      .string()
      .trim()
      .transform((initial, ctx) => {
        const result = initial.replace(/[-\s]/g, '')

        const digits: number[] = []
        for (const char of result) {
          const number = parseInt(char, 10)
          if (Number.isNaN(number)) {
            ctx.addIssue({
              code: 'custom',
              message: 'Invalid BSB format. Must be only comprised of numbers.',
            })
            return z.NEVER
          }
          digits.push(number)
        }

        if (digits.length !== 6) {
          ctx.addIssue({
            code: 'custom',
            message: 'Invalid BSB format. Must be 6 digits long.',
          })
          return z.NEVER
        }

        return digits.join('')
      }),
}) {
  static fakeInstance() {
    return BSB.fromJSON('123-456')
  }
}
