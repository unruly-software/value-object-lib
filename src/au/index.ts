export * from './abn'
export * from './acn'
export * from './bsb'
export * from './tfn'
