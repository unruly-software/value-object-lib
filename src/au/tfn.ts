import { ValueObject } from '@unruly-software/value-object'
import z from 'zod'

const TFNSchema = (weights: number[], length: number) =>
  z
    .string({
      message: 'Invalid TFN format. Must be a text value.',
    })
    .trim()
    .transform((initial, ctx) => {
      let cleanValue: string = ''
      const digits: number[] = []

      for (const char of initial) {
        /** These characters are commonly paired with TFN's */
        if (char === '-' || char === ' ' || char === '.' || char === '\n') {
          continue
        }
        const number = parseInt(char, 10)
        if (Number.isNaN(number)) {
          ctx.addIssue({
            code: z.ZodIssueCode.custom,
            message: 'Invalid TFN format. Must be a text value.',
          })
          return z.NEVER
        }
        cleanValue += char
        digits.push(number)
      }

      if (digits.length !== length) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: `Invalid TFN format. Must be ${length} digits long.`,
        })
        return z.NEVER
      }

      const checksum = digits.reduce((sum, number, index) => {
        const weightedNumber = number * weights[index]
        return sum + weightedNumber
      }, 0)

      if (checksum == 0 || checksum % 11 != 0) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: 'Invalid TFN checksum.',
        })
        return z.NEVER
      }

      return cleanValue
    })

export class IndividualTFN extends ValueObject.define({
  id: '@unruly-software/value-object-lib/IndividualTFN',
  schema: () => TFNSchema([1, 4, 3, 7, 5, 8, 6, 9, 10], 9),
}) {
  static fake() {
    return IndividualTFN.fromJSON('459599230')
  }
}

export class BusinessTFN extends ValueObject.define({
  id: '@unruly-software/value-object-lib/BusinessTFN',
  schema: () => TFNSchema([10, 7, 8, 4, 6, 3, 5, 1, 0], 8),
}) {
  static fake(): BusinessTFN {
    return new BusinessTFN('12345678')
  }
}

export class AnyTFN extends ValueObject.define({
  id: '@unruly-software/value-object-lib/AnyTFN',
  schema: () =>
    z.union([
      IndividualTFN.schema().transform((tfn) => tfn.props),
      BusinessTFN.schema().transform((tfn) => tfn.props),
    ]),
}) {
  getType(): 'IndividualTFN' | 'BusinessTFN' {
    if (this.props.length === 9) {
      return 'IndividualTFN'
    }
    if (this.props.length === 8) {
      return 'BusinessTFN'
    }
    throw new Error('Unexpected Error: Invalid validated TFN length')
  }
}
