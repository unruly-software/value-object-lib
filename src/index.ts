export * as au from './au'
