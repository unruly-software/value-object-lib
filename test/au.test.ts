import {describe, it, expect, expectTypeOf} from 'vitest'
import {ABN, ACN, AnyTFN, BSB, BusinessTFN, IndividualTFN} from '../src/au'
import z from 'zod'
import {ValueObject} from '@unruly-software/value-object'

const testValueObject = <T extends ValueObject.ValueObjectConstructor<string, any>>(ctor: T, tests: {
  invalid: unknown[]
  valid: [input: Parameters<T['fromJSON']>[0], output: InstanceType<T>['props']][]
}) => {
  describe(ctor[ValueObject.ValueObjectIdSymbol], () => {
    tests.invalid.forEach((input) => {
      it(`should reject invalid input: ${JSON.stringify(input)}`, () => {
        /** Very specifically using .safeParse to ensure zod schemas do not
         * throw outside of where expected */
        const parsed = ctor.schema().safeParse(input)

        expect(parsed.success).toBe(false)
        expect(parsed.error).toMatchSnapshot()
        expect(() => ctor.fromJSON(input)).toThrowError()
      })
    })

    tests.valid.forEach(([input, output]) => {
      it(`should accept valid input: ${JSON.stringify(input)}`, () => {
        const parsed = ctor.schema().safeParse(input)

        if (!parsed.success) {
          throw parsed.error
        }
        expect(parsed.data.props).toEqual(output)

        const instance = ctor.fromJSON(input)
        expect(instance.props).toEqual(output)
      })
    })
  })

}

describe('Australian Value Objects', () => {
  testValueObject(BSB, {
    valid: [['123 456', '123456'], ['123 - 457', '123457'], ['\n123\n456\n', '123456']],
    invalid: ['123', '123456789', 'no numbers', '1234567', '123-45a', '123-4567', '123-456-789', '123 and 456'],
  })

  testValueObject(ABN, {
    valid: [
      [
        '51 824 753 556',
        '51824753556',
      ],
      [
        '51-824-753-556',
        '51824753556',
      ],
      [
        '51 824 753 556',
        '51824753556',
      ],
      [
        ' 11 223 491 505 ',
        '11223491505',
      ],
      [
        '91 834 387 136',
        '91834387136',
      ],
      [
        ' 78 247 283 315',
        '78247283315',
      ],
    ],

    invalid: [
      'Hello world',
      '12345678912',
      'c26177aa-5b00-4808-b81a-34057bfc189e',
      '11834387136',
      '00000000000',
      '99999999999',
      'My ABN is 91 834 387 136'
    ]
  })

  testValueObject(ACN, {
    invalid: [
      'my acn is 000 000 019',
      'Hello world',
      '12345678912',
      'c26177aa-5b00-4808-b81a-34057bfc189e',
      '51 824 753 556',
      '51-824-753-556',
      '000 250 001',
      '000 500 004',
      '000 750 006',
      '001 000 003',
      '001 250 008',
      '001 500 002',
      '001 749 199',
    ],
    valid: [
      ['000 000 019', '000000019'],
      ['000-000-019', '000000019'],
      [' 000 000 019', '000000019'],
      ['000 250 000', '000250000'],
      ['000 500 005', '000500005'],
      ['000 750 005', '000750005'],
      ['001 000 004', '001000004'],
      ['001 250 004', '001250004'],
      ['001 500 009', '001500009'],
      ['001 749 999', '001749999'],
      ['001 999 999', '001999999'],
      ['002 249 998', '002249998'],
      ['002 499 998', '002499998'],
      ['002 749 993', '002749993'],
      ['002 999 993', '002999993'],
      ['003 249 992', '003249992'],
      ['003 499 992', '003499992'],
      ['003 749 988', '003749988'],
      ['003 999 988', '003999988'],
      ['004 249 987', '004249987'],
      ['004 499 987', '004499987'],
      ['004 749 982', '004749982'],
      ['004 999 982', '004999982'],
      ['005 249 981', '005249981'],
      ['005 499 981', '005499981'],
      ['005 749 986', '005749986'],
      ['005 999 977', '005999977'],
      ['006 249 976', '006249976'],
      ['006 499 976', '006499976'],
      ['006 749 980', '006749980'],
      ['006 999 980', '006999980'],
      ['007 249 989', '007249989'],
      ['007 499 989', '007499989'],
      ['007 749 975', '007749975'],
      ['007 999 975', '007999975'],
      ['008 249 974', '008249974'],
      ['008 499 974', '008499974'],
      ['008 749 979', '008749979'],
      ['008 999 979', '008999979'],
      ['009 249 969', '009249969'],
      ['009 499 969', '009499969'],
      ['009 749 964', '009749964'],
      ['009 999 964', '009999964'],
      ['010 249 966', '010249966'],
      ['010 499 966', '010499966'],
      ['010 749 961', '010749961'],
      ['\n010 749 961\n', '010749961'],
      ['010-749-961', '010749961'],
    ]
  })

  const individualTFNs: [string, string][] = [
    ['\n865414088\n', '865414088'],
    ['459-599-230', '459599230'],
    ['1124-740-82', '112474082'],
    ['565.051.603', '565051603'],
    ['907 974 668', '907974668'],
  ]

  const businessTFNs: [string, string][] = [
    ['81 854 402', '81854402',],
    ['37 118 629', '37118629',],
    ['37 118 660', '37118660',],
    ['37 118 705', '37118705',],
    ['38 593 474', '38593474',],
    ['38 593 519', '38593519',],
    ['85 655 734', '85655734',],
    ['85 655 797', '85655797',],
    ['85 655 810', '85655810',],
    ['37 118 655', '37118655'],
  ]

  const sharedTFNCases = [
    'not a number',
    'ooooooooo',
    '123456789',
    '37 118 656',
    'Longform prose is not expected to be valid',
  ]

  testValueObject(IndividualTFN, {
    invalid: [
      ...businessTFNs.map(([input]) => input),
      ...sharedTFNCases,
    ],
    valid: individualTFNs,
  })

  testValueObject(BusinessTFN, {
    invalid: [
      ...individualTFNs.map(([input]) => input),
      ...sharedTFNCases,
    ],
    valid: businessTFNs,
  })

  testValueObject(AnyTFN, {
    invalid: [
      ...sharedTFNCases,
    ],
    valid: [
      ...individualTFNs,
      ...businessTFNs,
    ],
  })
})
