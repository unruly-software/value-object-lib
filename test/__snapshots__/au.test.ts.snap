// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Australian Value Objects > @unruly-software/value-object-lib/ABN > should reject invalid input: "00000000000" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid ABN checksum.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/ABN > should reject invalid input: "11834387136" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid ABN checksum.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/ABN > should reject invalid input: "12345678912" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid ABN checksum.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/ABN > should reject invalid input: "99999999999" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid ABN checksum.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/ABN > should reject invalid input: "Hello world" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid ABN. Must be a text value.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/ABN > should reject invalid input: "My ABN is 91 834 387 136" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid ABN. Must be a text value.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/ABN > should reject invalid input: "c26177aa-5b00-4808-b81a-34057bfc189e" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid ABN. Must be a text value.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/ACN > should reject invalid input: "000 250 001" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid ACN checksum.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/ACN > should reject invalid input: "000 500 004" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid ACN checksum.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/ACN > should reject invalid input: "000 750 006" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid ACN checksum.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/ACN > should reject invalid input: "001 000 003" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid ACN checksum.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/ACN > should reject invalid input: "001 250 008" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid ACN checksum.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/ACN > should reject invalid input: "001 500 002" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid ACN checksum.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/ACN > should reject invalid input: "001 749 199" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid ACN checksum.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/ACN > should reject invalid input: "51 824 753 556" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid ACN format. Must be 11 digits long.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/ACN > should reject invalid input: "51-824-753-556" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid ACN format. Must be 11 digits long.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/ACN > should reject invalid input: "12345678912" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid ACN format. Must be 11 digits long.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/ACN > should reject invalid input: "Hello world" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid ACN. Must be a text value.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/ACN > should reject invalid input: "c26177aa-5b00-4808-b81a-34057bfc189e" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid ACN. Must be a text value.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/ACN > should reject invalid input: "my acn is 000 000 019" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid ACN. Must be a text value.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/AnyTFN > should reject invalid input: "37 118 656" 1`] = `
[ZodError: [
  {
    "code": "invalid_union",
    "errors": [
      [
        {
          "code": "custom",
          "message": "Invalid TFN format. Must be 9 digits long.",
          "path": []
        }
      ],
      [
        {
          "code": "custom",
          "message": "Invalid TFN checksum.",
          "path": []
        }
      ]
    ],
    "path": [],
    "message": "Invalid input"
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/AnyTFN > should reject invalid input: "123456789" 1`] = `
[ZodError: [
  {
    "code": "invalid_union",
    "errors": [
      [
        {
          "code": "custom",
          "message": "Invalid TFN checksum.",
          "path": []
        }
      ],
      [
        {
          "code": "custom",
          "message": "Invalid TFN format. Must be 8 digits long.",
          "path": []
        }
      ]
    ],
    "path": [],
    "message": "Invalid input"
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/AnyTFN > should reject invalid input: "Longform prose is not expected to be valid" 1`] = `
[ZodError: [
  {
    "code": "invalid_union",
    "errors": [
      [
        {
          "code": "custom",
          "message": "Invalid TFN format. Must be a text value.",
          "path": []
        }
      ],
      [
        {
          "code": "custom",
          "message": "Invalid TFN format. Must be a text value.",
          "path": []
        }
      ]
    ],
    "path": [],
    "message": "Invalid input"
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/AnyTFN > should reject invalid input: "not a number" 1`] = `
[ZodError: [
  {
    "code": "invalid_union",
    "errors": [
      [
        {
          "code": "custom",
          "message": "Invalid TFN format. Must be a text value.",
          "path": []
        }
      ],
      [
        {
          "code": "custom",
          "message": "Invalid TFN format. Must be a text value.",
          "path": []
        }
      ]
    ],
    "path": [],
    "message": "Invalid input"
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/AnyTFN > should reject invalid input: "ooooooooo" 1`] = `
[ZodError: [
  {
    "code": "invalid_union",
    "errors": [
      [
        {
          "code": "custom",
          "message": "Invalid TFN format. Must be a text value.",
          "path": []
        }
      ],
      [
        {
          "code": "custom",
          "message": "Invalid TFN format. Must be a text value.",
          "path": []
        }
      ]
    ],
    "path": [],
    "message": "Invalid input"
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/BSB > should reject invalid input: "123 and 456" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid BSB format. Must be only comprised of numbers.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/BSB > should reject invalid input: "123" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid BSB format. Must be 6 digits long.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/BSB > should reject invalid input: "123-45a" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid BSB format. Must be only comprised of numbers.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/BSB > should reject invalid input: "123-456-789" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid BSB format. Must be 6 digits long.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/BSB > should reject invalid input: "123-4567" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid BSB format. Must be 6 digits long.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/BSB > should reject invalid input: "1234567" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid BSB format. Must be 6 digits long.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/BSB > should reject invalid input: "123456789" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid BSB format. Must be 6 digits long.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/BSB > should reject invalid input: "no numbers" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid BSB format. Must be only comprised of numbers.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/BusinessTFN > should reject invalid input: "\\n865414088\\n" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid TFN format. Must be 8 digits long.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/BusinessTFN > should reject invalid input: "37 118 656" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid TFN checksum.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/BusinessTFN > should reject invalid input: "459-599-230" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid TFN format. Must be 8 digits long.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/BusinessTFN > should reject invalid input: "565.051.603" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid TFN format. Must be 8 digits long.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/BusinessTFN > should reject invalid input: "907 974 668" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid TFN format. Must be 8 digits long.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/BusinessTFN > should reject invalid input: "1124-740-82" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid TFN format. Must be 8 digits long.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/BusinessTFN > should reject invalid input: "123456789" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid TFN format. Must be 8 digits long.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/BusinessTFN > should reject invalid input: "Longform prose is not expected to be valid" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid TFN format. Must be a text value.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/BusinessTFN > should reject invalid input: "not a number" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid TFN format. Must be a text value.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/BusinessTFN > should reject invalid input: "ooooooooo" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid TFN format. Must be a text value.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/IndividualTFN > should reject invalid input: "37 118 629" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid TFN format. Must be 9 digits long.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/IndividualTFN > should reject invalid input: "37 118 655" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid TFN format. Must be 9 digits long.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/IndividualTFN > should reject invalid input: "37 118 656" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid TFN format. Must be 9 digits long.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/IndividualTFN > should reject invalid input: "37 118 660" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid TFN format. Must be 9 digits long.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/IndividualTFN > should reject invalid input: "37 118 705" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid TFN format. Must be 9 digits long.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/IndividualTFN > should reject invalid input: "38 593 474" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid TFN format. Must be 9 digits long.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/IndividualTFN > should reject invalid input: "38 593 519" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid TFN format. Must be 9 digits long.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/IndividualTFN > should reject invalid input: "81 854 402" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid TFN format. Must be 9 digits long.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/IndividualTFN > should reject invalid input: "85 655 734" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid TFN format. Must be 9 digits long.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/IndividualTFN > should reject invalid input: "85 655 797" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid TFN format. Must be 9 digits long.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/IndividualTFN > should reject invalid input: "85 655 810" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid TFN format. Must be 9 digits long.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/IndividualTFN > should reject invalid input: "123456789" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid TFN checksum.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/IndividualTFN > should reject invalid input: "Longform prose is not expected to be valid" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid TFN format. Must be a text value.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/IndividualTFN > should reject invalid input: "not a number" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid TFN format. Must be a text value.",
    "path": []
  }
]]
`;

exports[`Australian Value Objects > @unruly-software/value-object-lib/IndividualTFN > should reject invalid input: "ooooooooo" 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Invalid TFN format. Must be a text value.",
    "path": []
  }
]]
`;
